### ISSUE TITLE
a question of model training

### ISSUE THREAD

**Issue by muyangren123321** [dl4to/dl4to#9]
Hello, author. I have a question regarding the training of neural networks that I would like to ask you. In the topology optimization dataset you created, is there a difference in the problem definitions between consecutive frames? My understanding is that the main difference between the problem of frame 10 and the problem of frame 11 is the different values of problem.Ω_design, while other data remain the same. I have tried to verify this, and indeed problem.Ω_design is different, but I do not understand why problem.F is also different. My thought is that if my neural network inputs the problem data from frame 10, then the prediction of my neural network would be the density distribution for frame 11. Assuming this density value is predicted correctly, it should actually correspond to the value of problem.Ω_design.
```
dataset_train = SELTODataset(r'data/dataset', 'disc_simple', train=True)
dataset_val = SELTODataset(r'data/dataset', 'disc_simple', train=False)
dataloader_train = get_dataloader(dataset_train, batch_size=64, shuffle=True)
dataloader_val = get_dataloader(dataset_val, batch_size=64, shuffle=True)
problem, gt_solution = dataset_train[20]
problem1, gt_solution1 = dataset_train[50]
print(torch.equal(problem.Ω_dirichlet, problem1.Ω_dirichlet))
print(torch.equal(problem.Ω_design, problem1.Ω_design))
print(torch.equal(problem.F, problem1.F))

Found 1509 files.
importing dataset...
100%|██████████| 1509/1509 [00:00<00:00, 1960.53it/s]
done!
Found 200 files.
importing dataset...
100%|██████████| 200/200 [00:00<00:00, 2172.54it/s]
done!
True
False
False
```

**Comment by muyangren123321** [comment]
In addition, assuming my conjecture is correct, I have another question: the obtained solution's solution.θ is not entirely composed of 0, 1, or -1. How can I process solution.θ to convert it to 0, 1, or -1 without introducing significant errors?

**Comment by muyangren123321** [comment]
like this?
```
def transform_tensor(θ):
    tensor = θ.detach().numpy()
    tensor = np.where(tensor > 0.5, 1, tensor)
    tensor = np.where(tensor < -0.5, -1, tensor)
    tensor = np.where((tensor <= 0.5) & (tensor >= -0.5), 0, tensor)
    return torch.tensor(tensor)

**Comment by dl4to** [comment]
Regarding your first question:
Please take a look at our paper "SELTO: Sample-Efficient Learned Topology Optimization" (https://arxiv.org/pdf/2209.05098). On the bottom on page 4, we write that the both datasets (disc and sphere) "[...] have fixed Dirichlet boundary conditions but variable force positions and magnitudes.". This is the reason why in your example code, `problem.Ω_dirichlet` and `problem1.Ω_dirichlet` are the same.  However, the forces are different in each sample, that is why `problem.F` and `problem1.F` are different. Regarding the design space: Please have a look on page 5 of the aforementioned paper. Here we write: "For voxels that have Dirichlet boundary conditions or loads assigned to them we enforce the density value to be 1 by setting `Ω_design = 1`." Here, "loads" refers to the forces `F`. This means: At voxels with non-zero forces, we fix the density to be one. That is the reason why `problem.Ω_design` and `problem1.Ω_design` are different (because the forces given by `problem.F` and `problem1.F` are different). I hope that answers the question!

Regarding your second question:
The density values should always be between 0 and 1, and never negative (0 means that the voxel is empty, and 1 means that it is completely full of material). You can use the `solution.get_θ(binary=True)` to get a thresholded version of the density, with the threshold being 0.5. Note that `solution.get_θ(binary=False)` would be the same as simply `solution.θ`.

**Comment by muyangren123321** [comment]
Thank you for your answer. I now fully understand your explanation, which means that the problems in your created database are entirely different, and the neural network you trained directly predicts the final optimization results from the initial conditions. Regarding the second question, I would like to add something to my approach. My idea is whether it's possible to implement iterative optimization for a single topology optimization problem based on the problem and solution data format you created. This is a reinforcement learning problem, where the state is input, an action is taken, and the state is updated. For example, if my neural network inputs the problem data from frame 10, that serves as my state. The network then predicts a density value, and I should update my state for the next prediction, similar to the SIMP algorithm. In the 10th iteration step of the algorithm, the density value output should serve as the relevant input for the 11th iteration step. Therefore, my question is whether I can update the problem's Ω_design with the density value predicted in the previous step, whether through the algorithm or the model, but I need to process the predicted density value, just like in the code you provided. Specifically, assign all predicted density values greater than 0.5 to 1, those less than -0.5 to -1, and set those between -0.5 and 0.5 to 0.

**Comment by dl4to** [comment]
I am not entirely sure I understand your approach, so let me sum up the idea to check if I understand you correctly:

you want to train a reinforcement learning (RL) model that gets as input a state (Ω_design, Ω_dirichlet, F), and outputs θ (in RL logic, this is the "action"). Then, based on this density θ, you want to update the state for the next iteration, and you want to achieve this by modifying Ω_design. Correct so far? 
Further, you would like to "stretch" the original definition of the density θ and allow density values between -1 and 1. Here, I still do not understand why you would want that. How do you interpret a density of -1? Do you scale it back to [0,1] in the end, to have meaningful densities?

Let's repeat what the design space Ω_design expresses; A value of 1 means that the corresponding density at that voxel is always fixed at 1, a value of 0 means that it is always fixed at 0, and a value of -1 means that this density is free to be optimized and has no external limitation. So if you set the design space to 1 for densities greater than 0.5, then these densities will be fixed at a value of 1. Conversely, if you set the design space to 0 for densities between -0.5 and 0.5, then the densities in these voxels will be set to and fixed at 0. Then only the densities between -0.5 and -1 remain how they are. Maybe this is what you want, but I currently do not see how this is useful. And: Wouldn't is be possible to achieve the same thing if you leave the densities between 0 and 1 (as they were before), and set the design space to 1 if θ>0.66, set it to 0 if 0.33<θ<0.66, and to -1 if θ<0.33? That would be more in line with the library and with the meaning/interpretation of densities.

On a sidenote: Currently, in the library it is not intended to change Ω_design in a problem, after the problem has been created. So, in order to come up with a problem with a different Ω_design, I would create a new one:
problem_new = Problem(E=problem.E, ..., Ω_dirichlet=Ω_dirichlet_new, Ω_design=problem.Ω_dirichlet, ...)

I do not know if this is helpful, but I wrote a few sentences on reinforcement learning in my PhD thesis. Please take a look at Section 4.4.2 here: https://media.suub.uni-bremen.de/handle/elib/8405

**Comment by muyangren123321** [comment]
Thank you for your response. I believe that the way you mentioned updating Ω_design is correct. I made a mistake by thinking that the density values are distributed between -1 and 1, which was a basic error. Inspired by your explanation, I wonder if it would be appropriate to set the positions in Ω_design to 0 where the predicted θ is less than 0.33, effectively removing those voxels, while keeping Ω_design unchanged elsewhere. This would result in a new 3D voxel model, which represents the updated state based on the previous action.
Additionally, since the direction and position of the forces in the dataset you created are different, it means that the dataset actually contains n distinct design problems.
Once again, thank you for your answer and assistance.
